pkg/collector/corechecks/servicediscovery/module/impl_linux.go

```diff
@@ -886,26 +881,72 @@ func getServiceNameFromContainerTags(tags []string) (string, string) {
 	return "", ""
 }
 
-func (s *discovery) enrichContainerData(service *model.Service, containers map[string]*agentPayload.Container, pidToCid map[int]string) {
-	id, ok := pidToCid[service.PID]
+func (s *discovery) getContainersMap() map[int]*workloadmeta.Container {
+	containers := s.wmeta.ListContainersWithFilter(workloadmeta.GetRunningContainers)
+	containersMap := make(map[int]*workloadmeta.Container, len(containers))
+
+	metricsProvider := metrics.GetProvider(option.New(s.wmeta))
+
+	for _, container := range containers {
+		collector := metricsProvider.GetCollector(provider.NewRuntimeMetadata(
+			string(container.Runtime),
+			string(container.RuntimeFlavor),
+		))
+		if collector == nil {
+			containersMap[int(container.PID)] = container
+			continue
+		}
+
+		pids, err := collector.GetPIDs(container.Namespace, container.ID, containerCacheValidity)
+		if err != nil || len(pids) == 0 {
+			containersMap[int(container.PID)] = container
+			continue
+		}
+
+		for _, pid := range pids {
+			containersMap[int(pid)] = container
+		}
+	}
+	return containersMap
+}
+
+func (s *discovery) getProcessContainerInfo(pid int, containers map[int]*workloadmeta.Container) (string, []string, bool) {
+	container, ok := containers[pid]
 	if !ok {
-		return
+		return "", nil, false
+	}
+
+	containerID := container.EntityID.ID
+	collectorTags := container.CollectorTags
+
+	// Getting the tags from the tagger. This logic is borrowed from
+	// the GetContainers helper in pkg/process/util/containers.
+	entityID := types.NewEntityID(types.ContainerID, containerID)
+	entityTags, err := s.tagger.Tag(entityID, types.HighCardinality)
+	if err != nil {
+		log.Tracef("Could not get tags for container %s: %v", containerID, err)
+		return containerID, collectorTags, false
 	}
+	tags := append(collectorTags, entityTags...)
+
+	return containerID, tags, true
+}
 
-	container, ok := containers[id]
+func (s *discovery) enrichContainerData(service *model.Service, containers map[int]*workloadmeta.Container) {
+	containerID, containerTags, ok := s.getProcessContainerInfo(service.PID, containers)
```

Comment:
If there are multiple services per container, then the tag collection will be done several times for the same container, maybe you can consider if you should cache it?